coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/UnionSqlCallTransformer.java

```diff
@@ -0,0 +1,165 @@
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+
+import com.linkedin.coral.common.transformers.SqlCallTransformer;
+import com.linkedin.coral.common.utils.TypeDerivationUtil;
+
+import static org.apache.calcite.rel.rel2sql.SqlImplementor.POS;
+import static org.apache.calcite.sql.parser.SqlParserPos.ZERO;
+
+
+/**
+ * This transformer class is used to adapt expressions in
+ * set statements (e.g.: `UNION`, `INTERSECT`, `MINUS`)
+ * in case that the branches of the set statement contain fields in
+ * char family which have different types.
+ * The `char` fields which are differing from the expected `varchar` output
+ * of the set statement will be adapted through an explicit `CAST` to the `varchar` type.
+ *
+ * @see <a href="https://github.com/trinodb/trino/issues/9031">Change char/varchar coercion in Trino</a>
+ */
+public class UnionSqlCallTransformer extends SqlCallTransformer {
+
+  public UnionSqlCallTransformer(TypeDerivationUtil typeDerivationUtil) {
+    super(typeDerivationUtil);
+  }
+
+  @Override
+  protected boolean condition(SqlCall sqlCall) {
+    return sqlCall.getOperator().kind == SqlKind.UNION || sqlCall.getOperator().kind == SqlKind.INTERSECT
+        || sqlCall.getOperator().kind == SqlKind.MINUS;
+  }
+
+  @Override
+  protected SqlCall transform(SqlCall sqlCall) {
+    List<SqlNode> operandsList = sqlCall.getOperandList();
+    List<List<Optional<RelDataType>>> columnsTypesLists = new ArrayList<>();
+    for (SqlNode operand : operandsList) {
+      if (operand.getKind() == SqlKind.SELECT && ((SqlSelect) operand).getSelectList() != null) {
+        SqlSelect select = (SqlSelect) operand;
+        List<SqlNode> selectNodes = select.getSelectList().getList();
+        if (columnsTypesLists.isEmpty()) {
+          for (int i = 0; i < selectNodes.size(); i++) {
+            columnsTypesLists.add(new ArrayList<>());
+          }
+        }
+
+        for (int i = 0; i < selectNodes.size(); i++) {
+          SqlNode sqlNode = selectNodes.get(i);
+          if (sqlNode.getKind() == SqlKind.IDENTIFIER && ((SqlIdentifier) sqlNode).isStar()) {
+            // The type derivation of * on the SqlNode layer is not supported now.
+            return sqlCall;
+          }
+
+          Optional<RelDataType> selectNodeDataType;
+          try {
+            selectNodeDataType = Optional.of(deriveRelDatatype(sqlNode));
+          } catch (RuntimeException e) {
+            // The type derivation may fail for complex expressions
+            selectNodeDataType = Optional.empty();
+          }
+          columnsTypesLists.get(i).add(selectNodeDataType);
+        }
+      } else {
+        return sqlCall;
+      }
+    }
+
+    List<SqlNode> updatedOperands = new ArrayList<>(sqlCall.getOperandList().size());
+    for (SqlNode operand : operandsList) {
```

Comment:
Any reason we iterate on the operandsList twice? Can we merge this loop and the above? I see below we cast to SqlSelect in the first loop we check if it is SqlSelect. In both we iterate on the selectNodes, etc. I think there is room for consolidation? PS. In any case let us reuse the naming conventions suggested above.

Reply:
We need to find the least restrictive type for each select item index from the operands - this is one looping.

Once we have the least restrictive type, we need another looping to verify whether the operands need to be updated.

I didn't find a better alternative to do this.